### PR TITLE
fix: Add model validation for reference files in init command

### PR DESCRIPTION
**Pull request by beatrizsavinhas**
## Description

Validation is not performed for the reference file models when running the init command. If incorrect fields are set in the `REFERENCE_FILES` dict, there is no information communicated to the user and the error is propagated further along the logic.

Bug discovered while working on https://github.com/Clinical-Genomics/BALSAMIC/pull/1651.

#### Added

- model validation for reference pydantic models `ReferencesCanFam` and `ReferencesHg`
- `model_config = ConfigDict(extra="forbid")` to both models to warn the user with an error in case extra unknown arguments are provided
- `test_init_CanFam`
- `test_get_reference_output_paths_canfam`

## Documentation

- [x] N/A

### Feature Tests

- [x] Provide an additional argument in the `REFERENCE_FILES` dict - in this example `cytoband_coordinates_file`, without updating the attributes for `ReferencesHg`.
	image

- [x] Check that an error is thrown when running the init command:
	image

- [x] Make sure all reference files can be downloaded as before.

### Pipeline Integrity Tests

- **Report deliver (generation of the `.hk` file)**
    - [x] N/A
- **TGA T/O Workflow**
    - [x] N/A
- **TGA T/N Workflow**
    - [x] N/A
- **UMI T/O Workflow**
    - [x] N/A
- **UMI T/N Workflow**
    - [x] N/A
- **WGS T/O Workflow**
    - [x] N/A
- **WGS T/N Workflow**
    - [x] N/A
- **QC Workflow**
    - [x] N/A
- **PON Workflow**
    - [x] N/A

### Documentation

- **Atlas documentation**
    - [x] N/A
- **Web portal for Clinical Genomics**
    - [x] N/A

### Panel of Normal specific criteria

- [ ] The PR includes the addition of a new Panel of Normals
- [ ] The samples have been verified to adhere to the sample selection criteria on [Atlas PoN creation instructions for Balsamic](https://atlas.scilifelab.se/infrastructure/BALSAMIC/balsamic_pon/#procedure)

### User Changes

- [x] N/A - No changes in any files or pipeline functionality except extra error warnings in case of manual changes done by the developers.

### Infrastructure Changes

- **Stored files in Housekeeper**
    - [x] N/A    - [ ] Updated: [Link]
- **CG (CLI and delivered/uploaded files)**
    - [x] N/A
- **Servers (configuration files on Hasta)**
    - [x] N/A
- **Scout interface**
    - [x] N/A

### Validation criteria

Validation criteria to be added to validation report PR: [LINK-TO-VALIDATION-REPORT-PR from the [validations repository](https://github.com/Clinical-Genomics/validations/)]

**Version specific criteria**

-  N/A

> [!IMPORTANT]
> One of the below checkboxes for validation need to be checked 

- [ ] Added version specific validation criteria to validation report
- [ ] Changes validated in standard sections: [validation-section]
- [X] Validation criteria not necessary

## Checklist

> [!IMPORTANT]  
> Ensure that all checkboxes below are ticked before merging.

### For Developers

- **PR Description**
    - [x] Provided a comprehensive description of the PR.
    - [x] Linked relevant user stories or issues to the PR.
- **Documentation**
    - [x] Verified and updated documentation if necessary.
- **Validation criteria**
    - [x] Completed the validation criteria section of the template.
- **Tests**
    - [x] Described and tested the functionality addressed in the PR.
    - [x] Ensured integration of the new code with existing workflows.
    - [x] Confirmed that meaningful unit tests were added for the changes introduced.
    - [x] Checked that the PR has successfully passed all relevant code smells and coverage checks.
- **Review**
    - [x] Addressed and resolved all the feedback provided during the code review process.
    - [x] Obtained final approval from designated reviewers.

### For Reviewers

- **Code**
    - [ ] Code implements the intended features or fixes the reported issue.
    - [ ] Code follows the project's coding standards and style guide.
- **Documentation**
    - [ ] Pipeline changes are well-documented in the CHANGELOG and relevant documentation.
- **Validation criteria**
    - [ ] The author has completed the validation criteria section of the template
- **Tests**
    - [ ] The author provided a description of their manual testing, including consideration of edge cases and boundary
      conditions where applicable, with satisfactory results.
- **Review**
    - [ ] Confirmed that the developer has addressed all the comments during the code review.

### CODE REVIEW

File: BALSAMIC/models/cache.py [hunk 1]
```diff
@@ -145,6 +148,8 @@ class ReferencesHg(References):
     vcf_1kg: ReferenceUrl
     wgs_calling_regions: ReferenceUrl
 
+    model_config = ConfigDict(extra="forbid")
```
Comment (mathiasbio) [inline BALSAMIC/models/cache.py]: Would it work to add this to the References base model and have it be inherited? Not important, just curious!

File: BALSAMIC/models/cache.py [hunk 2]
```diff
@@ -298,7 +303,7 @@ class CacheConfig(BaseModel):
     cosmic_key: Optional[str] = None
     bioinfo_tools: dict
     containers: Dict[str, str]
-    references: Union[ReferencesHg, ReferencesCanFam]
+    references: ReferencesHg | ReferencesCanFam
```
Comment (mathiasbio) [inline BALSAMIC/models/cache.py]: I wonder how this works with the new `model_config = ConfigDict(extra="forbid")` will it be used again when initialising the model? And if so, how would it know which Reference model to use? 🤔

**Review by mathiasbio — APPROVED** [review]
Looks like a nice improvement ⭐ I think adding the canfam test is fine if it helps us get rid of the code coverage warning. The alternative is cleaning up canfam entirely I guess, and that's not worth the time-investment probably! 

One final thing for me is just reassurance that the regular balsamic config case is still working. From what I could tell -- while the reference dict is validated in the config case model, it doesn't seem to validate based on any specific genome version, so I think it's fine. And if you merge I can verify that it's working in one of my other branches as I'm already running tests : )

File: BALSAMIC/models/cache.py (hunk 2 above)
Comment (mathiasbio) [reply BALSAMIC/models/cache.py]: I mean this section where the model is initialised:

```
cache_config: CacheConfig = CacheConfig(
        analysis={"case_id": f"reference.{genome_version}.{cache_version}"},
        references_dir=references_dir.as_posix(),
        genome_dir=genome_dir.as_posix(),
        variants_dir=variants_dir.as_posix(),
        vep_dir=vep_dir.as_posix(),
        containers_dir=containers_dir.as_posix(),
        genome_version=genome_version,
        cosmic_key=cosmic_key,
        bioinfo_tools=BIOINFO_TOOL_ENV,
        containers=get_containers(cache_version),
        references=references,
        references_date=datetime.now().strftime("%Y-%m-%d %H:%M"),
    )
```

This is validating the fields in the model which includes: 

`references: Union[ReferencesHg, ReferencesCanFam] `or `references: ReferencesHg | ReferencesCanFam` in this version. I'm quite sure that this isn't simply a typehint, but actually initialising that nested reference model as well, which would activate the validation 🤔 but I could be wrong

**Comment by beatrizsavinhas** [comment]
SonarCloud issues addressed! ✅ 

codecov points to coverage issues in files/code that was not touched in this PR so it's out of scope to create tests for that.
Screenshot 2026-02-06 at 12 57 14
